src/ParamPassing.cc

```diff
@@ -0,0 +1,675 @@
+          + std::string(_childrenXml->Attribute("element_id")) + "'>:\n"
+          + ElementToString(xmlChild)
+        });
+        continue;
+      }
+    }
+
+    ElementPtr elemChild;
+    if (useParentDesc)
+      elemChild = elemDesc;
+    else
+      elemChild = elemDesc->GetElementDescription(elemName);
+
+    if (!readXml(xmlChild, elemChild, _errors))
+    {
+      _errors.push_back({ErrorCode::ELEMENT_INVALID,
+        "Unable to convert XML to SDF. Skipping child element alteration "
+        "with parent <" + std::string(_childrenXml->Name()) + " element_id='"
+        + std::string(_childrenXml->Attribute("element_id")) + "'>:\n"
+        + ElementToString(xmlChild)
+      });
+      continue;
+    }
+
+    if (actionStr == "add")
+    {
+      _elem->InsertElement(elemChild);
+    }
+    else if (actionStr == "replace")
+    {
+      ElementPtr e = getElementByName(_elem, xmlChild);
+      if (e == nullptr)
+      {
+        _errors.push_back({ErrorCode::ELEMENT_MISSING,
+          "Could not find element. Skipping child element replacement "
+          "with parent <" + std::string(_childrenXml->Name()) + " element_id='"
+          + std::string(_childrenXml->Attribute("element_id")) + "'>:\n"
+          + ElementToString(xmlChild)
+        });
+        continue;
+      }
+
+      // check name requirement of original and replacement element
+      if (e->HasAttribute("name") && !xmlChild->Attribute("name"))
+      {
+        _errors.push_back({ErrorCode::ATTRIBUTE_MISSING,
+          "Replacement element is missing a 'name' attribute. "
+          "Skipping element replacement <" + std::string(_childrenXml->Name())
+          + " element_id='" + std::string(_childrenXml->Attribute("element_id"))
+          + "'>:\n" + ElementToString(xmlChild)
+        });
+        continue;
+      }
+
+      replace(elemChild, e);
+    }
+  }
+}
+
+
+//////////////////////////////////////////////////
+void add(tinyxml2::XMLElement *_childXml, ElementPtr _elem, Errors &_errors)
+{
+  ElementPtr newElem = initElementDescription(_childXml, _errors);
+  if (!newElem)
+    return;
+
+  if (readXml(_childXml, newElem, _errors))
+  {
+    _elem->InsertElement(newElem);
+  }
+  else
+  {
+    _errors.push_back({ErrorCode::ELEMENT_INVALID,
+      "Unable to convert XML to SDF. Skipping element addition:\n"
+      + ElementToString(_childXml)
+    });
+  }
+}
+
+//////////////////////////////////////////////////
+void modifyAttributes(tinyxml2::XMLElement *_xml,
+                      ElementPtr _elem, Errors &_errors)
+{
+  for (const tinyxml2::XMLAttribute *attr = _xml->FirstAttribute();
+      attr; attr = attr->Next())
+  {
+    std::string attrName = attr->Name();
+
+    if (attrName == "element_id" || attrName == "action")
+      continue;
+
+    if (!_elem->HasAttribute(attrName))
+    {
+      // TODO(jenn) add attributes for namespaced elements?
+      if (attrName.find(":") != std::string::npos)
```

Comment:
This is needed because I'm reading the attributes from tinyxml (listed in `<experimental:params>`) and adding the attribute to the original element if there isn't one or it's a custom attribute (not using functions from parser.cc). For example, if a user wanted to add an attribute `test` this would fail, but if they used `custom:test` it would succeed. Should I change it to skip the `_elem->HasAttribute` check and add the attribute(s) anyway?